test/State.hs

```diff
@@ -67,3 +76,28 @@ test m a s (Run runState) =
   , testProperty "put updates the state variable" . forall (s :. s :. m a :. Nil) $
     \ s s' m -> runState (s, put s' >> m) === runState (s', m)
   ]
+
+problematic :: Applicative f => f a
+problematic = (pure (error "should immediately")) <*> pure (error "terminate")
+
+lazy
+  :: (Monad m, Functor f)
+  => GenTerm (f ())
+  -> Run f g m
+  -> [TestTree]
+lazy s (Run run) =
+  [ testProperty "fmap is non-strict" . forall (a :. s :. Nil) $
+    \ a s -> void . eval . run $ (const a <$> pure (error "insufficiently lazy")) <$ s
```

Comment:
`eval`, from Hedgehog, ensures that evaluating the passed term doesn’t raise an exception.